src/otx/v2/adapters/torch/dataset.py

```diff
@@ -0,0 +1,199 @@
+"""OTX adapters.torch.mmengine.mmpretrain.Dataset API."""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Iterable
+
+import yaml
+from torch.utils.data import DataLoader as TorchDataLoader
+from torch.utils.data import Dataset as TorchDataset
+from torch.utils.data import Sampler
+
+from otx.v2.api.core.dataset import BaseDataset
+from otx.v2.api.utils import set_tuple_constructor
+from otx.v2.api.utils.decorators import add_subset_dataloader
+
+SUBSET_LIST = ["train", "val", "test"]
+
+
+@add_subset_dataloader(SUBSET_LIST)
+class TorchBaseDataset(BaseDataset):
+    """A class representing a dataset for pretraining a model."""
+
+    def _initialize(self) -> None:
+        self.set_datumaro_adapters()  # Set self.dataset_entity & self.label_schema
+        self.initialize = True
+
+    def build_dataset(
+        self,
+        subset: str,
+        pipeline: list | dict | None = None,
+        config: str | dict | None = None,
+    ) -> TorchDataset | None:
+        """Builds a TorchDataset object for the given subset using the specified pipeline and configuration.
+
+        Args:
+            subset (str): The subset to build the dataset for.
+            pipeline (Optional[Union[list, dict]]): The pipeline to use for the dataset.
+                Defaults to None.
+            config (Optional[Union[str, dict]]): The configuration to use for the dataset.
+                Defaults to None.
+
+        Returns:
+            Optional[TorchDataset]: The built TorchDataset object, or None if the dataset is empty.
+        """
+        raise NotImplementedError
+
+
+    def build_dataloader(
```

Comment:
I am not sure why this is required. This is because, dataset is built per each subset level and each subset is going to generate its own dataloader. So, from my guess, I don't think this is used ever. So my suggestion is removing this and rename `subset_dataloader` to `build_dataloader`. What do you think?

Reply:
Hi @wonjuleee, I'll answer this with a comment on the subset_dataloader function above.

Currently, `otx.v2.api.core.dataset::BaseDataset` consists of the functions `build_dataset`, `build_dataloader`, and `subset_dataloader`.
`build_dataset` is a function that creates a torch dataset based on the input,
`build_dataloader` creates a torch Dataloader based on the torch Dataset.
`subset_dataloader` is a convenience function that calls build_dataset and build_dataloader internally.
Using the subset_dataloader function and a decorator function called add_subset_dataloader, we can provide a call to .train_dataloader(), val_dataloader(), test_dataloader() from this dataset object, which behaves internally as subset_dataloader(subset="train").

We might consider renaming subset_dataloader to build_subset_dataloader. @samet-akcay Any thoughts?

Reply:
It's a common sense spread out widely for those who use `torch.data.Dataset` and `torch.data.DataLoader` that `torch.data.Dataset` includes only a single subset and there exist a 1:1 corresponded `torch.data.DataLoader` to it. In this view, `build_dataset` implies `build_subset_of_dataset` and `build_dataloader` does too for many practitioners. For me, this seems some duplications.

Reply:
@wonjuleee @vinnamkim 
Yes, that's right, I agree.
`build_dataset` returns a torch Dataset with a subset as input,
`build_dataloader` returns a torch Dataloader with torch Dataset as input.
`subset_dataloader` returns a torch Datalaoder with subset as input.

We can use 
```
train_dataset = dataset.build_dataset(subset="train")
train_dataloader = dataset.build_dataloader(dataset=train_dataset)
```
but If we have a subset_dataloader function, then we can also get a dataloader in the same way as
```
train_dataloader = dataset.train_dataloader()
# OR train_dataloader = dataset.subset_dataloader(subset="train")
```
and so on.

It seems redundant, but I think it would be a nice function for users, what do you think?

Reply:
I am still not sure. I think users may want to have build_dataloader for each subset, and hence subset_dataloader can bring confusions. What do you think @vinnamkim?

Reply:
It is same story with my comment below.

1. `dataset.build_dataset()` returns `dataset` but it is actually the subset of `dataset`.
2. `dataset.build_dataloader()` can take any `dataset` (not just for itself) but anyway return `dataloader`, so that `dataloader` can have no relation with the `dataset` which created it.
3. On the other hand, `dataset.subset_dataloader()` can only produce `dataloader` which should produce a subset of `dataset`.

Let's get back to the step 2. Why do we need this kind of interface as a member function?

Reply:
I agree that `build_dataloader` feels redundant. I don't  think we need 3 different ways to obtain the dataloader of a given subset (in fact, I'm concerned it could cause confusion). To keep the terminology consistent I would be in favor of @wonjuleee's proposal to rename `subset_dataloader` to `build_dataloader` and remove the current `build_dataloader`. This way, both `build_dataset` and `build_dataloader` require passing the subset.

We could consider making `build_dataset` and `build_dataloader` private methods, to signal to the user that the only recommended way to obtain the dataloader of a given subset is the `<subset>_dataloader` method.

Reply:
Yes I think everyone's commands make sense. Thanks for the comment, I will proceed to replace `build_dataset` and `build_dataloader` with private functions.
And we'll discuss renamings for `subset_dataloader` a bit later.